File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <bits/stdc++.h>
using namespace std;

int row, col;

bool rangeCheck(int tmpRow, int tmpCol){
    if(0 <= tmpRow && tmpRow < row && 0 <= tmpCol && tmpCol < col) return true;
    return false;
}

bool roadCheck(int tmpRow, int tmpCol, vector<vector<int>>&maps){
    if(rangeCheck(tmpRow, tmpCol) && maps[tmpRow][tmpCol] == 1) return true;
    return false;
}

int solution(vector<vector<int>> maps)
{
    row = maps.size();
    col = maps[0].size();
    vector<vector<bool>>visited(row, vector<bool>(col, false));
    queue<tuple<int,int,int>>q;
    q.push({0, 0, 1});
    while(!q.empty()){
        int tmpRow = get<0>(q.front());
        int tmpCol = get<1>(q.front());
        int tmpCount = get<2>(q.front());
        if(tmpRow == row-1 && tmpCol == col-1) return tmpCount;
        q.pop();
        if(visited[tmpRow][tmpCol]) continue;
        visited[tmpRow][tmpCol] = true;
        if(roadCheck(tmpRow - 1, tmpCol, maps)) {
            q.push({tmpRow - 1, tmpCol, tmpCount + 1});
        }
        if(roadCheck(tmpRow + 1, tmpCol, maps)) {
            q.push({tmpRow + 1, tmpCol, tmpCount + 1});
        }
        if(roadCheck(tmpRow, tmpCol - 1, maps)) {
            q.push({tmpRow, tmpCol - 1, tmpCount + 1});
        }
        if(roadCheck(tmpRow, tmpCol + 1, maps)) {
            q.push({tmpRow, tmpCol + 1, tmpCount + 1});
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    int mapRow;
    int mapCol;
    int[][] map;
    
    boolean isDestination(int row, int col){
        if(row == mapRow-1 && col == mapCol-1) return true;
        return false;
    }
    
    boolean canMove(int row, int col){
        if(rangeCheck(row, col) && map[row][col] == 1) return true;
        return false;
    }
    
    boolean rangeCheck(int row, int col){
        if(0 <= row && row < mapRow && 0 <= col && col < mapCol) return true;
        return false;
    }
    
    class Pos{
        int row;
        int col;
        int count;
        Pos(int row, int col, int count){
            this.row = row;
            this.col = col;
            this.count = count;
        }
    }
    
    public int solution(int[][] maps) {
        map = maps;
        mapRow = maps.length;
        mapCol = maps[0].length;
        boolean[][] visited = new boolean[mapRow][mapCol];
        
        Queue<Pos> q = new LinkedList<>();
        q.add(new Pos(0, 0, 1));
        
        while(!q.isEmpty()){
            Pos tmp = q.poll();
            int row = tmp.row, col = tmp.col, count = tmp.count;
            
            if(visited[tmp.row][tmp.col]) continue;
            visited[tmp.row][tmp.col] = true;
            
            if(isDestination(row, col)) return count;
            
            if(canMove(row-1, col)) q.add(new Pos(row-1, col, count+1));
            if(canMove(row+1, col)) q.add(new Pos(row+1, col, count+1));
            if(canMove(row, col-1)) q.add(new Pos(row, col-1, count+1));
            if(canMove(row, col+1)) q.add(new Pos(row, col+1, count+1));
        }
        return -1;
    }
}